### PR TITLE
Adding logic for broker retry (StorageHelper keyswap fallback)

### PR DESCRIPTION
**Pull request by iambmelt**
Iterates on #534

### CODE REVIEW

File: common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java [hunk 1]
```diff
@@ -220,42 +231,77 @@ public String encrypt(final String clearText)
...
+            Logger.error(
+                    TAG + methodName,
+                    "Failed to decrypt with package name: "
+                            + packageName,
+                    e
+            );
+
+            if (!BROKER_PACKAGE_NAMES.contains(packageName)) {
+                // If we are not the broker, do not retry, immediately throw.
+                throw e;
```
Comment (sdabbiru) [inline common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java]: This check can be moved to right after line 250 (final String packageName = mContext.getPackageName();) - we should not even attempting to read if we are not one of these 2 apps.

File: common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java [hunk 2]
```diff
@@ -220,42 +231,77 @@ public String encrypt(final String clearText)
...
+            }
+
+            // We are the broker... try with the other broker's package name
+            final String nextPackageToUse =
+                    packageName.equals(AZURE_AUTHENTICATOR_APP_PACKAGE_NAME)
```
Comment (sdabbiru) [inline common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java]: >equals [](start = 32, length = 6)

Ignore case?

Comment (sdabbiru) [reply common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java]: Or does it matter?

---
In reply to: [296825941](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/535#discussion_r296825941) [](ancestors = 296825941)

File: common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java [hunk 3]
```diff
@@ -276,16 +322,47 @@ public String decrypt(final String encryptedBlob)
...
+                            encodeVersionLength
+                    )
+            );
+        }
+
+        if (!encryptedBlob.substring(1, 1 + encodeVersionLength).equals(ENCODE_VERSION)) {
+            throw new IllegalArgumentException(
+                    String.format(
+                            "Encode version received was: '%s', Encode version supported is: '%s'",
+                            encryptedBlob,
```
Comment (sdabbiru) [inline common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java]: >encryptedBlob [](start = 28, length = 13)

Should we be dumping this in the exception message?

File: common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java (hunk 1 above)
Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java]: Not true; this is shared code with MSAL

File: common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java (hunk 2 above)
Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java]: Package name [cannot change](https://android-developers.googleblog.com/2011/06/things-that-cannot-change.html) post-publish. All of our package names are lowercase, so makes no difference

File: common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java (hunk 3 above)
Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java]: We should not, will remove. Thanks

Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java]: Resolved in 05bfd4cc9398988713ed8f8b55e232aa6dc83ad9

**Comment by iambmelt** [comment]
@sdabbiru Edits made as requested/relevant

**Pull request merged**